Rubberduck.VBEEditor/Events/VbeEvents.cs

```diff
@@ -0,0 +1,227 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Rubberduck.VBEditor.SafeComWrappers;
+using Rubberduck.VBEditor.SafeComWrappers.Abstract;
+
+namespace Rubberduck.VBEditor.Events
+{
+    public sealed class VBEEvents : IVBEEvents
+    {
+        private static VBEEvents _instance;
+        private static readonly object Lock = new object();
+        private readonly IVBProjects _projects;
+        private readonly Dictionary<string, IVBComponents> _components;
+
+        public static VBEEvents Initialize(IVBE vbe)
+        {
+            lock (Lock)
+            {
+                if (_instance == null)
+                {
+                    _instance = new VBEEvents(vbe);
+                }
+            }
+
+            return _instance;
+        }
+
+        public static void Terminate()
+        {
+            lock (Lock)
+            {
+                if (_instance == null)
+                {
+                    return;
+                }
+
+                _instance.Dispose();
+                _instance = null;
+            }
+        }
+
+        private VBEEvents(IVBE vbe)
+        {
+            _components = new Dictionary<string, IVBComponents>();
+
+            if (_projects != null)
+            {
+                return;
+            }
+            
+            _projects = vbe.VBProjects;
+
+            _projects.AttachEvents();
+            _projects.ProjectAdded += ProjectAddedHandler;
+            _projects.ProjectRemoved += ProjectRemovedHandler;
+            _projects.ProjectRenamed += ProjectRenamedHandler;
+            _projects.ProjectActivated += ProjectActivatedHandler;
+            foreach (var project in _projects)
+            using (project)
+            {
+                {
+                    RegisterComponents(project);
+                }
+            }
+        }
+
+        private void RegisterComponents(IVBProject project)
+        {
+            if (project.IsWrappingNullReference || project.Protection != ProjectProtection.Unprotected)
+            {
+                return;    
+            }
+
+            if (string.IsNullOrWhiteSpace(project.ProjectId))
+            {
+                project.AssignProjectId();
+            }
+
+            var components = project.VBComponents;
+            _components.Add(project.ProjectId, components);
+
+            components.AttachEvents();
+            components.ComponentAdded += ComponentAddedHandler;
+            components.ComponentRemoved += ComponentRemovedHandler;
+            components.ComponentRenamed += ComponentRenamedHandler;
+            components.ComponentActivated += ComponentActivatedHandler;
+            components.ComponentSelected += ComponentSelectedHandler;
+            components.ComponentReloaded += ComponentReloadedHandler;
+        }
+
+        private void UnregisterComponents(string projectId)
+        {
+            if (!_components.ContainsKey(projectId))
```

Comment:
Why not `TryGet`? (It is on the `IDictionary` interface.)

Reply:
Did not know about that. Will update.

Reply:
Hm. On further reflection, I have concerns about potential leaks. If you look a bit further, we have a `using(var components = _components[projectId])`. If I take a `VBComponents` as a `out` parameter, I'm not sure if I can safely manage that variable without leak?